### PR TITLE
Fix padding of negative byte numbers in bitwise operation

### PR DESCRIPTION
**Pull request by DASPRiD**
Call me stupid: When padding the binaryX and Y values, I didn't pad with the individual sign byte but always with a positive sign byte. All new tests pass now.

### CODE REVIEW

File: src/Internal/Bitwise.php [hunk 1]
```diff
@@ -24,78 +24,33 @@
...
-        }
-
-        if ($x->isNegative()) {
-            $x2 = self::twosComplement($x2);
-        }
-
-        if ($y->isNegative()) {
-            $y2 = self::twosComplement($y2);
-        }
+        $binaryX = $binaryX[0] . str_pad(substr($binaryX, 1), $maxLength - 1, $binaryX[0], STR_PAD_LEFT);
```
Comment (DASPRiD) [inline src/Internal/Bitwise.php]: These two lines are the only change to my original implementation, the rest is just removing your bit-string prototype.

**Comment by DASPRiD** [comment]
I saw that you made some implementation changes to your prototype there. I'd still argue that my implementation is faster, as it requires far less conditional checks. What do you think about that?

**Comment by BenMorel** [comment]
Ah sorry, I missed your PR while I was working on an implementation myself!

I'm not sure about the speed, should we benchmark?

**Comment by DASPRiD** [comment]
I guess a benchmark wouldn't hurt :)

**Comment by BenMorel** [comment]
OK here is the benchmark in seconds, repeating 1000 times all the tests:

```
Yours: 10.810 10.489 10.544
Mine:   8.738  8.499  8.609
```

Wow I did not expect both of them to be that slow. I'm looking forward to using GMP when available. Anyway, there is a slightly better performance in my implementation at the moment.

Note that I did use the latest version of your code, using Calculator and full string bitwise operators. Your code uses indeed less conditional checks in the bitwise() method, but the same work has to be done anyway somewhere!

If you don't mind, I'll stick with my implementation, in addition to being marginally faster, it better matches my mental representation of things, and I will have to maintain this library, so... :-)

Anyway I learned a lot today, thanks for this!

**Comment by BenMorel** [comment]
I almost forgot to include the benchmark code:

```php
use Brick\Math\BigInteger;
use Brick\Math\Tests\BigIntegerTest;

require 'vendor/autoload.php';

$bigIntegerTest = new BigIntegerTest();

// pre-instantiate BigIntegers

$testsOr = [];
foreach ($bigIntegerTest->providerOr() as [$x, $y]) {
    $testsOr[] = [BigInteger::of($x), BigInteger::of($y)];
}

$testsAnd = [];
foreach ($bigIntegerTest->providerAnd() as [$x, $y]) {
    $testsAnd[] = [BigInteger::of($x), BigInteger::of($y)];
}
$testsXor = [];

foreach ($bigIntegerTest->providerXor() as [$x, $y]) {
    $testsXor[] = [BigInteger::of($x), BigInteger::of($y)];
}

$t = microtime(true);

for ($i = 0; $i < 1000; $i++) {
    foreach ($testsOr as [$a, $b]) {
        $a->or($b);
    }
    foreach ($testsAnd as [$a, $b]) {
        $a->and($b);
    }
    foreach ($testsXor as [$a, $b]) {
        $a->xor($b);
    }
}

printf("%.3f\n", microtime(true) - $t);
```

**Comment by DASPRiD** [comment]
Sure, you can stick with your implementation. I'd still say that it could be optimized in some places, like using str_pad() instead of str_repeat(), but that's probably a micro-optimization. Personally I just prefer to use existing functionality when it exists :)

Looking forward to version 0.8 anyway, so I can continue with my project :)

**Pull request closed without merging**